autorest/azure/devicetoken.go

```diff
@@ -0,0 +1,196 @@
+	req, _ := autorest.Prepare(
+		&http.Request{},
+		autorest.AsPost(),
+		autorest.WithBaseURL(OAuthDeviceEndpoint),
+		autorest.WithFormData(url.Values{
+			"client_id": []string{clientID},
+			"resource":  []string{resource},
+		}),
+		autorest.WithQueryParameters(map[string]interface{}{
+			authAPIVersionQueryParamName: authAPIVersionQueryParamValue,
+		}))
+
+	resp, err := client.Send(req)
+	if err != nil {
+		return nil, fmt.Errorf("Error occurred while requesting Device Authorization Code. %s", err)
+	}
+	if resp.StatusCode != http.StatusOK {
+		return nil, fmt.Errorf(
+			"Request for Device Authorization Code resulted in Status: %d: %s",
+			resp.StatusCode,
+			http.StatusText(resp.StatusCode))
+	}
+
+	var code DeviceCode
+	decoder := json.NewDecoder(resp.Body)
+	err = decoder.Decode(&code)
+	if err != nil {
+		return nil, fmt.Errorf("Failed to decode Device Authorization Code json")
+	}
+
+	code.Resource = &resource
+
+	return &code, nil
+}
+
+// CheckForUserCompletion takes a DeviceCode and checks with the Azure AD OAuth endpoint
+// to see if the device flow has: been completed, timed out, or otherwise failed
+func CheckForUserCompletion(client *autorest.Client, clientID string, code *DeviceCode) (*Token, error) {
+	req, _ := autorest.Prepare(
+		&http.Request{},
+		autorest.AsPost(),
+		autorest.WithBaseURL(OAuthTokenEndpoint),
+		autorest.WithFormData(url.Values{
+			"client_id":  []string{clientID},
+			"code":       []string{*code.DeviceCode},
+			"grant_type": []string{OAuthGrantTypeDeviceCode},
+			"resource":   []string{*code.Resource},
+		}),
+		autorest.WithQueryParameters(map[string]interface{}{
+			authAPIVersionQueryParamName: authAPIVersionQueryParamValue,
+		}))
+
+	resp, err := client.Send(req)
+	if err != nil {
+		return nil, fmt.Errorf("Error occurred while exchanging device code for token. %s", err)
+	}
+	if resp.StatusCode != http.StatusOK && resp.StatusCode != http.StatusBadRequest {
+		return nil, fmt.Errorf(
+			"Exchanging Device Code for OAuth token resulted in Status: %d: %s",
+			resp.StatusCode,
+			http.StatusText(resp.StatusCode))
+	}
+
+	var token DeviceToken
+	decoder := json.NewDecoder(resp.Body)
+	err = decoder.Decode(&token)
+	if err != nil {
+		return nil, fmt.Errorf("Failed to decode Device Token json")
+	}
+
+	if token.Error == nil {
+		return &token.Token, nil
+	}
+
+	switch *token.Error {
+	case "authorization_pending":
+		return nil, ErrDeviceAuthorizationPending
+	case "slow_down":
+		return nil, ErrDeviceSlowDown
+	case "access_denied":
+		return nil, ErrDeviceAccessDenied
+	case "code_expired":
+		return nil, ErrDeviceCodeExpired
+	default:
+		return nil, ErrDeviceGeneric
+	}
+}
+
+// WaitForUserCompletion calls CheckForUserCompletion repeatedly until a token is granted or an error state occurs.
+// This prevents the user from looping and checking against 'ErrDeviceAuthorizationPending'.
+func WaitForUserCompletion(client *autorest.Client, clientID string, code *DeviceCode) (*Token, error) {
```

Comment:
Per your other comment and writing a bit of it, I'm not going to implement this for now. One, Azure times out anyway. Two, I don't think it's undesirable to loop forever, if anything, there would be a user configurable timeout that would be passed in and respected. Otherwise the user can just call CheckFor... and do whatever they want.